File: OfficialCode/polynomialls.cpp
#include<stdio.h>
#include<math.h>
#define n 9
#define m 2
void gauss(double a[m+1][m+1],double b[m+1],double x[m+1])
{      double l[m+1][m+1];
		double sum;
		

	  int i,j,k;
	 for (k=0;k<=m-1;k++)
{                                                           
     for (i=k+1;i<=m;i++)
     {
        l[i][k]=a[i][k]/a[k][k];
         b[i]=b[i]-l[i][k]*b[k];
		 
         for (j=k+1;j<=m;j++)
            {
           a[i][j]=a[i][j]-l[i][k]*a[k][j];
		  
             }
     }
}

x[m]=b[m]/a[m][m];

 for (k=m-1;k>=0;k--)
{
    sum=0;
   for (j=k+1;j<=m;j++)
    {
    sum=sum+a[k][j]*x[j];
    }  
   
     x[k]=(b[k]-sum)/a[k][k];
	
}
 
}
int main()
{
  int i,j,k,l;
	double G[m+1][m+1],d[m+1],a[m+1];
	double sum,error=0.,ls;
	
	double x[n+1]={0.,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9};
	double y[n+1]={0.,5.1234,5.3057,5.5680,5.9378,6.4370,7.0978,7.9493,9.0253,10.3627};
	double w[n+1]={0.,1.,1.,1.,1.,1.,1.,1.,1.,1.};
	printf("正则方程组的系数矩阵\n"); 
	for (k=0;k<=m;k++)
	{
		for (j=0;j<=m;j++)
		{
			G[k][j]=0.;
			for (i=1;i<=n;i++)
			{
				G[k][j]=G[k][j]+w[i]*pow(x[i],k+j);
			}
			printf("%lf\t",G[k][j]);
		}
		printf("\n");
	}
	printf("正则方程组的右端向量\n");
	for (j=0;j<=m;j++)
	{
		d[j]=0.;
		for (i=1;i<=n;i++)
		{
			d[j]=d[j]+w[i]*y[i]*pow(x[i],j);
		}
		printf("%lf\n",d[j]);
	}

	gauss(G,d,a);
	printf("\n拟合的二次多项式\n");
	for (i=0;i<=m;i++)
	{
		if (a[i]<0)
		printf("(%lf)",a[i]);
		else
         printf("%lf",a[i]);
		for (j=1;j<i;j++)
		{
			printf("x*");
		}
		if (j==i)
			printf("x");
		if (i==m)
			break;
		printf("+");

	}
	for (i=1;i<=n;i++)
	{   ls=0.;
		for (l=0;l<=m;l++)
		{
			ls=ls+a[l]*pow(x[i],l);
		}
		error=error+w[i]*pow(ls-y[i],2.);
	}
	error=sqrt(error);
	//printf("\n均方误差%e\n",error);
	
  return 0;
}

File: OfficialCode/polynomialls.m
n=9;
m=2;
G=zeros(m+1,m+1);
d=zeros(1,m+1);
a=zeros(1,m+1);
x=[0.,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9];
y=[0.,5.1234,5.3057,5.5680,5.9378,6.4370,7.0978,7.9493,9.0253,10.3627];
w=[0.,1.,1.,1.,1.,1.,1.,1.,1.,1.]; %便于计算，第一个元素不考虑

fprintf('正则方程组的系数矩阵');
for k=0:m
    for j=0:m
        G(k+1,j+1)=0;
        for i=1:n
            G(k+1,j+1)=G(k+1,j+1)+w(i+1)*x(i+1).^(k+j);
        end
    end
end
G

fprintf('正则方程组的右端向量');
for j=0:m
    d(j+1)=0;
    for i=1:n
        d(j+1)=d(j+1)+w(i+1)*y(i+1)*x(i+1).^j;
    end
end
d

a=gauss(G,d,m);
fprintf('拟合的二次多项式');
for i=0:m
    if a(i+1)<0
        fprintf('(%d)',a(i+1));
    else
        fprintf('%d',a(i+1));
    end
    j=1;
    if j<=i
    for j=1:i-1
        fprintf('x*');
    end
    j=j+1;
    end
    if j==i
        fprintf('x');
    end
    if i==m
        break;
    end
         fprintf('+');
end

function [x]=gauss(a,b,m)
for k=0:m-1
    for i=k+1:m
        l(i+1,k+1)=a(i+1,k+1)/a(k+1,k+1);
        b(i+1)=b(i+1)-l(i+1,k+1)*b(k+1);
        for j=k+1:m
            a(i+1,j+1)=a(i+1,j+1)-l(i+1,k+1)*a(k+1,j+1);
        end
    end
end
x(m+1)=b(m+1)/a(m+1,m+1);
for k=m-1:-1:0
    sum=0;
    for j=k+1:m
        sum=sum+a(k+1,j+1)*x(j+1);
    end
    x(k+1)=(b(k+1)-sum)/a(k+1,k+1);
end
end
